### PR TITLE
feat: Add metrics input schemas and database helpers (PR 6a)

### PR DESCRIPTION
**Pull request by harsh04044**
**What kind of change does this PR introduce?**

Feature - Adds GraphQL input schemas and database helper utilities for metrics operations

**Issue Number:**

Fixes #4125 

**Snapshots/Videos:**

N/A - This PR adds backend schema definitions and utility functions. No UI changes.

**If relevant, did you update the documentation?**

Yes - Documentation has been automatically generated:
- TypeDoc documentation for all utility functions in `docs/docs/auto-docs/utilities/metrics/dbHelpers/`
- GraphQL schema documentation for new input types in `docs/docs/auto-schema/types/inputs/`

**Summary**

This PR implements the foundation for metrics querying functionality by introducing GraphQL input schemas and reusable database helper utilities. The implementation provides type-safe interfaces for querying metrics data with flexible filtering options, and includes statistical calculation utilities that will be used for metrics analysis.

The changes include:
- **MetricsInput**: Schema for basic metrics queries with time range filtering, operation type filtering, and duration-based filtering
- **MetricsFilterInput**: Advanced filtering schema supporting operation name filtering, slow operations detection, cache hit rate filtering, and complexity range filtering
- **dbHelpers**: Utility functions for statistical calculations (mean, median, percentile) and time window generation for metrics aggregation

This work is part of PR 6a and sets the groundwork for future metrics querying and analysis features. All code follows existing project patterns, includes comprehensive JSDoc documentation, and maintains 100% test coverage.

**Does this PR introduce a breaking change?**

No - This is a purely additive change. New GraphQL input types are added without modifying any existing functionality.

### CodeRabbit AI Review
- [ ] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [ ] I have implemented or provided justification for each non-critical suggestion
- [ ] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [x] I have written tests for all new changes/features
- [x] I have verified that test coverage meets or exceeds 95%
- [x] I have run the test suite locally and all tests pass

**Other information**

- Used `isNotNullish` utility for consistency with existing codebase patterns
- Removed non-null assertions in favor of proper bounds checking to comply with linter rules
- All functions are pure (no side effects) and well-documented
- Comprehensive edge case handling in all utility functions
- Rebased onto latest upstream develop branch

**Implementation Details:**

**New Files Created:**
- `src/graphql/inputs/MetricsInput.ts` - Basic metrics query input schema
- `src/graphql/inputs/MetricsFilterInput.ts` - Advanced metrics filtering input schema (includes nested ComplexityRangeInput)
- `src/utilities/metrics/dbHelpers.ts` - Statistical and time window utility functions
- `test/graphql/inputs/MetricsInput.test.ts` - 39 test cases
- `test/graphql/inputs/MetricsFilterInput.test.ts` - 45 test cases  
- `test/utilities/metrics/dbHelpers.test.ts` - 76 test cases

**Modified Files:**
- `src/graphql/inputs/index.ts` - Added exports for new input schemas

**Test Coverage:**
- 160+ test cases covering all code paths
- 100% patch coverage for all new code
- All edge cases tested (empty arrays, null/undefined, boundary values, error conditions)

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

## Summary by CodeRabbit

* **New Features**
  * Added three GraphQL input types for complexity ranges, advanced metrics filtering, and metrics queries (time range, duration bounds, operation type, cache inclusion) with cross-field validation and sensible defaults.

* **Utilities**
  * New metrics helpers for mean, median, percentile, time-window generation, and time-range normalization with robust validation and immutability.

* **Tests**
  * Extensive unit tests covering new inputs, utilities, validation rules, and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by harsh04044** [comment]
The `black --check` failure is an upstream CI issue. The file [.github/workflows/scripts/check_docstrings.py](cci:7://file:///Users/harshchandwani/Desktop/gsoc/talawa-api/.github/workflows/scripts/check_docstrings.py:0:0-0:0) is identical to `upstream/develop` and passes black locally (v25.11.0).

The CI's unpinned `black` in [.github/workflows/requirements.txt](cci:7://file:///Users/harshchandwani/Desktop/gsoc/talawa-api/.github/workflows/requirements.txt:0:0-0:0) may be installing a different version. @palisadoes   help resolve this

File: src/utilities/metrics/dbHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+export function getTimeWindows(
+	startTime: Date,
+	endTime: Date,
+	options?: TimeWindowOptions,
+): TimeWindow[] {
+	if (startTime >= endTime) {
+		throw new Error(
+			`startTime must be before endTime. startTime: ${startTime.toISOString()}, endTime: ${endTime.toISOString()}`,
+		);
+	}
```
[the remark this answers is not in the document]
Comment (harsh04044) [reply src/utilities/metrics/dbHelpers.ts:266]: @coderabbitai this is resolved

**Comment by harsh04044** [comment]
@coderabbitai Full review

**Comment by harsh04044** [comment]
@coderabbitai Full review

**Comment by harsh04044** [comment]
@coderabbitai Full review

File: src/graphql/inputs/MetricsInput.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+			isNotNullish(arg.minDuration) &&
+			isNotNullish(arg.maxDuration) &&
+			arg.minDuration >= arg.maxDuration
+		) {
+			ctx.addIssue({
+				code: "custom",
+				message: `minDuration must be less than maxDuration. minDuration: ${arg.minDuration}, maxDuration: ${arg.maxDuration}`,
+				path: ["minDuration"],
+			});
+		}
```
[the remark this answers is not in the document]
Comment (harsh04044) [reply src/graphql/inputs/MetricsInput.ts:57]: this is resolved

**Comment by harsh04044** [comment]
@coderabbitai Full review

File: test/graphql/inputs/MetricsFilterInput.test.ts [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+ * 3. GraphQL integration tests will be added when resolvers are implemented to test end-to-end
+ *
+ * TODO: Add mercuriusClient integration tests when GraphQL resolvers are implemented (PR 6b).
+ * These tests should:
+ * - Create GraphQL document nodes for queries/mutations accepting MetricsFilterInput
+ * - Test operationNames validation through GraphQL layer
+ * - Test complexityRange min/max validation through GraphQL layer
+ * - Verify error messages contain proper validation details
+ * @see https://github.com/PalisadoesFoundation/talawa-api/issues/XXXX (follow-up issue)
+ */
```
[the remark this answers is not in the document]
Comment (harsh04044) [reply test/graphql/inputs/MetricsFilterInput.test.ts:25]: @coderabbitai Full review

**Comment by harsh04044** [comment]
@coderabbitai Full review

**Comment by harsh04044** [comment]
@coderabbitai Full review

**Comment by harsh04044** [comment]
@coderabbitai post your analysis as comment

**Comment by harsh04044** [comment]
@palisadoes review this
